doc/sphinx/language/gallina-specification-language.rst

```diff
@@ -779,7 +779,55 @@ type which is the type of its body.
 A formal presentation of constants and environments is given in
 Section :ref:`typing-rules`.
 
-.. cmd:: Definition @ident := @term
+.. cmd:: @def_token @ident_decl @def_body
+   :name: Definition; Example; SubClass
+
+   .. insertprodn def_token pattern_occ
+
+   .. prodn::
+      def_token ::= Definition
+      | Example
+      | SubClass
+      def_body ::= {* @binder } := {? @reduce } @term
+      | {* @binder } : @term := {? @reduce } @term
+      | {* @binder } : @term
+      reduce ::= Eval @red_expr in
+      red_expr ::= red
```

Comment:
I'll break out the grammar for red_expr separately.

I don't know precisely where the dividing line between the kernel and the "elaboration engine" lies.

> I'm also a bit worried how much the objective of describing the complete syntax of commands at once conflicts with the refactoring objective of first describing the kernel, then the elaboration engine

Reality is messier than our neat conceptions.  The refactoring is not yet started.  (Indeed, we've not even put out the CEP plan for doing so--which I hope we can do within, say, the next month.)

A useful way to think about **any** decision is how difficult/impactful it would be to reverse.  If it's difficult to do so then you should do more analysis up front.  If it's easy, then much less analysis.  Seems to me it would be quite easy to move a `red_expr` subsection between chapters.

Reply:
If you want `red_expr` elsewhere I will put it there now.

Reply:
For now, let's have a new subsection "Computation" say between "Assertions" and "Attributes" where you can move the grammar from `reduce` to `pattern_occ`.